### PR TITLE
feat(route): fix syosetu & add more routes

### PR DESCRIPTION
**Pull request by SnowAgar25**
## Involved Issue / 该 PR 相关 Issue

Close #

## Example for the Proposed Route(s) / 路由地址示例

```routes
/syosetu/n1976ey?limit=2
/syosetu/n5589gn?limit=2
/syosetu/search/yomou/word=陽キャ&notword=&genre=&type=&mintime=&maxtime=&minlen=50000&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/search/yomou/word=陽キャ&notword=&genre=101-401&type=&mintime=&maxtime=&minlen=50000&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/search/noc/word=ハーレム&notword=娼館&type=r&mintime=&maxtime=&minlen=30000&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&isgl=1&notbl=1&order=new?limit=5
/syosetu/search/mnlt/word=聖女&notword=&nocgenre=2&type=&mintime=&maxtime=&minlen=&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/search/mnlt/word=聖女&notword=&nocgenre=3&type=&mintime=&maxtime=&minlen=&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/search/mnlt/word=聖女&notword=&nocgenre=&type=&mintime=&maxtime=&minlen=&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/search/mid/word=学園&notword=&type=&mintime=&maxtime=&minlen=50000&maxlen=&min_globalpoint=&max_globalpoint=&minlastup=&maxlastup=&minfirstup=&maxfirstup=&order=new?limit=5
/syosetu/rankingr18/noc/daily_total?limit=5
/syosetu/rankingr18/mnlt/daily_total?limit=5
/syosetu/rankingr18/mnlt-bl/daily_total?limit=5
/syosetu/rankingr18/mid/daily_total?limit=5
/syosetu/rankingr18/noc/daily_t?limit=5
/syosetu/rankingr18/mnlt/daily_t?limit=5
/syosetu/rankingr18/mnlt-bl/daily_t?limit=5
/syosetu/rankingr18/mid/daily_t?limit=5
/syosetu/ranking/list/daily_total?limit=5
/syosetu/ranking/list/total_total?limit=5
/syosetu/ranking/genre/daily_101?limit=5
/syosetu/ranking/isekai/daily_2?limit=5
/syosetu/ranking/list/daily_t?limit=5
/syosetu/ranking/list/total_t?limit=5
/syosetu/ranking/genre/daily_101_t?limit=5
/syosetu/ranking/isekai/daily_2_t?limit=5
/syosetu/dev
```

## New RSS Route Checklist / 新 RSS 路由检查表
  
- [x] New Route / 新的路由
  - [x] Follows [Script Standard](https://docs.rsshub.app/joinus/advanced/script-standard) / 跟随 [路由规范](https://docs.rsshub.app/zh/joinus/advanced/script-standard)
- [ ] Anti-bot or rate limit / 反爬/频率限制
  - [ ] If yes, do your code reflect this sign? / 如果有, 是否有对应的措施?
- [x] [Date and time](https://docs.rsshub.app/joinus/advanced/pub-date) / [日期和时间](https://docs.rsshub.app/zh/joinus/advanced/pub-date)
  - [x] Parsed / 可以解析
  - [x] Correct time zone / 时区正确
- [x] New package added / 添加了新的包
- [ ] `Puppeteer`

### CODE REVIEW

File: lib/routes/syosetu/dev.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            source: ['dev.syosetu.com'],
+            target: '/dev',
+        },
+    ],
+};
+
+async function handler(): Promise<Data> {
+    const url = 'https://dev.syosetu.com';
+
+    return (await cache.tryGet(url, async () => {
```
Comment (TonyRL) [inline lib/routes/syosetu/dev.ts:34]: This is not suggested to be cached.

File: lib/routes/syosetu/index.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            source: ['novel18.syosetu.com/:ncode', 'ncode.syosetu.com/:ncode'],
+            target: '/:ncode',
+        },
+    ],
+};
+
+async function handler(ctx: Context): Promise<Data> {
+    const { ncode } = ctx.req.param();
+    const routeParams = querystring.parse(ctx.req.param('routeParams'));
+    const limit = Math.min(Number(routeParams.limit || 5), 20);
```
Comment (TonyRL) [inline lib/routes/syosetu/index.ts:38]: Please make use of the [common parameter limit](https://docs.rsshub.app/guide/parameters#limit-entries) `ctx.req.query('limit')` instead.

File: lib/routes/syosetu/ranking-r18.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+async function handler(ctx: Context): Promise<Data> {
+    const { sub, type } = ctx.req.param();
+    const baseUrl = `https://${sub === 'mnlt-bl' ? 'mnlt' : sub}.syosetu.com`;
+    const rankingUrl = `${baseUrl}/rank/list/type/${type}`;
+    const api = new NarouNovelFetch();
+
+    const routeParams = querystring.parse(ctx.req.param('routeParams'));
+    const limit = Number(routeParams.limit || 300);
```
Comment (TonyRL) [inline lib/routes/syosetu/ranking-r18.ts:206]: Please make use of the [common parameter limit](https://docs.rsshub.app/guide/parameters#limit-entries) `ctx.req.query('limit')` instead.

File: lib/routes/syosetu/ranking-r18.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            case SyosetuSub.MIDNIGHT:
+                return R18Site.Midnight;
+            default:
+                throw new InvalidParameterError(`Invalid subsite: ${sub}`);
+        }
+    })();
+
+    const builder = new SearchBuilderR18(searchParams, api).r18Site(nocgenre);
+
+    const items = (await cache.tryGet(rankingUrl, async () => {
```
Comment (TonyRL) [inline lib/routes/syosetu/ranking-r18.ts:238]: This is not suggested to be cached.

File: lib/routes/syosetu/search.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    return new SearchBuilderR18(r18Params, new NarouNovelFetch());
+}
+
+async function handler(ctx: Context): Promise<Data> {
+    const { sub, query } = ctx.req.param();
+    const searchUrl = `https://${sub}.syosetu.com/search/search/search.php?${query}`;
+
+    return (await cache.tryGet(searchUrl, async () => {
```
Comment (TonyRL) [inline lib/routes/syosetu/search.ts:255]: Please make use of the [common parameter limit](https://docs.rsshub.app/guide/parameters#limit-entries) `ctx.req.query('limit')` instead.

File: lib/routes/syosetu/ranking.ts [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    return { period, category, novelType };
+}
+
+async function handler(ctx: Context): Promise<Data> {
+    const { listType, type } = ctx.req.param();
+    const rankingType = listType as RankingType;
+    const routeParams = querystring.parse(ctx.req.param('routeParams'));
+    const limit = Number(routeParams.limit || 300);
```
Comment (TonyRL) [inline lib/routes/syosetu/ranking.ts:285]: Please make use of the [common parameter limit](https://docs.rsshub.app/guide/parameters#limit-entries) `ctx.req.query('limit')` instead.

File: lib/routes/syosetu/ranking.ts [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+            };
+        }
+
+        default:
+            throw new InvalidParameterError(`Invalid ranking type: ${type}`);
+    }
+
+    const builder = new SearchBuilder(searchParams, api);
+
+    const items = (await cache.tryGet(rankingUrl, async () => {
```
Comment (TonyRL) [inline lib/routes/syosetu/ranking.ts:405]: This is not suggested to be cached.

**Comment by SnowAgar25** [comment]
@TonyRL Using ctx.req.query('limit') might have an issue.
Cache middleware only uses request path as key, which means first request's limit will determine what gets cached.
For example, if first request uses limit=5 (from getBest5RadarItems), later requests for limit=100 will still only return 5 items.
This could be a problem since I'm using getBest5 to provide top-5 listings - if these get cached first, users won't be able to get more items in their requests.
https://github.com/DIYgod/RSSHub/blob/d1dd8a53ca99b56a3075c9d3e587addf6ae6d49e/lib/middleware/cache.ts#L20

**Comment by TonyRL** [comment]
> Cache middleware only uses request path as key

#17674 updated to include limit as well

**Pull request merged**